### PR TITLE
Add GET Repository High Level REST API

### PR DESCRIPTION
**Pull request by hub-cap**
This commit adds the Modules Client with a first API call within it, the
get repositories call in snapshot/restore module. The test itself is
somewhat lacking, but that is due to the fact that there is no way to
insert more repositories until a PUT call is fleshed out.

Relates #27205

### CODE REVIEW

File: client/rest-high-level/src/main/java/org/elasticsearch/client/ModulesClient.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+import java.io.IOException;
+
+import static java.util.Collections.emptySet;
+
+/**
+ * A wrapper for the {@link RestHighLevelClient} that provides methods for accessing the various Modules APIs.
+ * <p>
+ * See <a href="https://www.elastic.co/guide/en/elasticsearch/reference/current/modules.html">Modules APIs on elastic.co</a>
+ */
+public final class ModulesClient {
```
Comment (javanna) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/ModulesClient.java]: why do we need a separate client for modules? I thought that we need something like this for plugins, but users may not know that some of our stuff is isolate in modules, given that it's all shipped with the rest of the codebase. I would add these methods to the existing client, yet I may have missed previous discussions around this, not sure.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java [hunk 2]
```diff
@@ -252,6 +253,15 @@ public final ClusterClient cluster() {
         return clusterClient;
     }
 
+    /**
+     * Provides a {@link ModulesClient} which can be used to access the various Modules APIs.
+     *
+     * See <a href="https://www.elastic.co/guide/en/elasticsearch/reference/current/modules.html">Modules APIs on elastic.co</a>
+     */
+    public final ModulesClient modules() {
```
Comment (javanna) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java]: On the same line as my other comment above, I assume that this method is weird for users. Also, none of the other language clients have this namespace while they all have `indices`, `cluster` etc.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java [hunk 3]
```diff
@@ -630,6 +631,17 @@ static Request indexPutSettings(UpdateSettingsRequest updateSettingsRequest) thr
         return request;
     }
 
+    static Request modulesGetRepositories(GetRepositoriesRequest getRepositoriesRequest) {
```
Comment (javanna) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java]: could we add a test for this method?

File: client/rest-high-level/src/test/java/org/elasticsearch/client/ModulesClientIT.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+import static org.hamcrest.Matchers.equalTo;
+
+public class ModulesClientIT extends ESRestHighLevelClientTestCase {
+
+    public void testModulesGetRepositories() throws IOException {
+        GetRepositoriesRequest request = new GetRepositoriesRequest();
+        GetRepositoriesResponse response = execute(request, highLevelClient().modules()::getRepositories,
+            highLevelClient().modules()::getRepositoriesAsync);
+        // Once create repository is fleshed out, we will be able to do more with this test case.
```
Comment (javanna) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/ModulesClientIT.java]: couldn't we use the low-level REST client for this call? Maybe not worth it given that create is coming soon?

File: server/src/main/java/org/elasticsearch/action/admin/cluster/repositories/get/GetRepositoriesResponse.java [hunk 5]
```diff
@@ -74,4 +76,11 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
         builder.endObject();
         return builder;
     }
+
+    public static GetRepositoriesResponse fromXContent(XContentParser parser) throws IOException {
+        if (parser.nextToken() != XContentParser.Token.START_OBJECT) {
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/action/admin/cluster/repositories/get/GetRepositoriesResponse.java]: we could use ensureExpectedToken here? I even wonder if this check is needed, I don't think we do this kind of check consistently in our parsing code. Out of curiosity, what happens without it?

**Review by javanna — CHANGES_REQUESTED** [review]
I left a couple of comments, @hub-cap do you plan on adding docs later once all repositories API are done? I don't mind as long as we make sure that we don't release these without docs (it happened before unfortunately, that's why I worry)

File: server/src/main/java/org/elasticsearch/action/admin/cluster/repositories/get/GetRepositoriesResponse.java (hunk 5 above)
Comment (hub-cap) [reply server/src/main/java/org/elasticsearch/action/admin/cluster/repositories/get/GetRepositoriesResponse.java]: its possible this is a bug. I believe its because the toXContent above is calling builder.startObject, I dont know how we typically handle these. the code in `RepositoriesMetaData` checks for the tokens like this, so i went with that. Ill gladly change it given a better example.

**Comment by hub-cap** [comment]
Hey @javanna as this is my first API call, i wasn't aware of the docs. I will be sure to look it up and add them as I go, lest I forget them as well!

As for separating them out, I only did this because the separation I thought we were doing (there is a different docs section for clusters and indices, so i figured this should follow suit). Ill gladly add them to the client itself, rather than in a special client. Let me try again :)

File: server/src/main/java/org/elasticsearch/action/admin/cluster/repositories/get/GetRepositoriesResponse.java (hunk 5 above)
Comment (javanna) [reply server/src/main/java/org/elasticsearch/action/admin/cluster/repositories/get/GetRepositoriesResponse.java]: leave it, it certainly doesn't hurt. maybe use ensureExceptedToken though

**Comment by hub-cap** [comment]
I am putting this on hold until I can find out what some of the clients do WRT the snapshot namespace. I would like to be consistent as possible. This includes all doc work (which is not yet started, but will be part of this PR). I have addressed the other PR comments.

**Comment by hub-cap** [comment]
After chatting with some of the lang clients, I found out that they tend to use the namespace defined in the rest-api-spec. In this case its `snapshot.*`, so I went ahead and made a `SnapshotClient`.  I still need to do docs and a changelog entry :)

**Comment by hub-cap** [comment]
@javanna I have addressed the docs changes, as well as the PR requests. I have also run the docs build to make sure the high level rest client docs did not fail :)

**Comment by hub-cap** [comment]
@elasticmachine test it

**Comment by hub-cap** [comment]
@elasticmachine test this please

**Review by nik9000 — APPROVED** [review]
It looks fine to me.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java [hunk 6]
```diff
@@ -252,6 +255,16 @@ public final ClusterClient cluster() {
...
+    /**
+     * Provides a {@link SnapshotClient} which can be used to access the Snapshot API.
+     *
+     * See <a href="https://www.elastic.co/guide/en/elasticsearch/reference/current/modules-snapshots.html">Snapshot API on elastic.co</a>
+     */
+    public final SnapshotClient snapshot() {
+        return snapshotClient;
+    }
+
+
```
Comment (javanna) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java]: super important: one less empty line? :)

File: client/rest-high-level/src/main/java/org/elasticsearch/client/SnapshotClient.java [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+ * See <a href="https://www.elastic.co/guide/en/elasticsearch/reference/current/modules-snapshots.html">Snapshot API on elastic.co</a>
+ */
+public final class SnapshotClient {
+    private final RestHighLevelClient restHighLevelClient;
+
+    SnapshotClient(RestHighLevelClient restHighLevelClient) {
+        this.restHighLevelClient = restHighLevelClient;
+    }
+
+
```
Comment (javanna) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/SnapshotClient.java]: here too one less empty line?

File: client/rest-high-level/src/main/java/org/elasticsearch/client/SnapshotClient.java [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+    }
+
+    /**
+     * Asynchronously gets a list of snapshot repositories. If the list of repositories is empty or it contains a single element "_all", all
+     * registered repositories are returned.
+     * <p>
+     * See <a href="https://www.elastic.co/guide/en/elasticsearch/reference/current/modules-snapshots.html"> Snapshot and Restore
+     * API on elastic.co</a>
+     */
+
```
Comment (javanna) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/SnapshotClient.java]: remove this empty line?

File: client/rest-high-level/src/test/java/org/elasticsearch/client/RequestConvertersTests.java [hunk 9]
```diff
@@ -1427,6 +1428,24 @@ public void testIndexPutSettings() throws IOException {
         assertEquals(expectedParams, request.getParameters());
     }
 
+    public void testGetRepositories() {
+        Map<String, String> expectedParams = new HashMap<>();
+        expectedParams.put("master_timeout", "30s");
```
Comment (javanna) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/RequestConvertersTests.java]: you can use setRandomTimeout. also can we test the local flag (you can use setRandomLocal I think)

**Review by javanna — APPROVED** [review]
left a few small comments, LGTM otherwise, feel free to merge once you addressed those. Thanks!

File: client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java [hunk 10]
```diff
@@ -792,6 +803,13 @@ Params withRefreshPolicy(WriteRequest.RefreshPolicy refreshPolicy) {
             return this;
         }
 
+        Params withRepository(String[] repositories) {
+            if (repositories != null && repositories.length > 0) {
+                return putParam("repository", Strings.arrayToCommaDelimitedString(repositories));
```
[the remark this answers is not in the document]
Comment (hub-cap) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java]: yes, ive fixed this. Good call.

File: client/rest-high-level/src/test/java/org/elasticsearch/client/RequestConvertersTests.java (hunk 9 above)
Comment (hub-cap) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/RequestConvertersTests.java]: thats awesome, i fixed the master timeout and local params by using these helpers. tyvm for showing them to me.

**Comment by hub-cap** [comment]
@elasticmachine test this please
